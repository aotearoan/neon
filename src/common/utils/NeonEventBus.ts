/**
 * Event bus used internally for component to component communication. All Neon messages are namespaced with the
 * messagePrefix below. Feel free to also use it in your applications, just use your own namespace so as not to clash
 * with Neon's messages.
 */
export class NeonEventBus {
  public static messagePrefix = 'neon-';

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private static listeners: Record<string, Array<(...args: any[]) => void>> = {};

  /**
   * Listen to event bus events
   * @param event {string}
   * @param callback {(...args: any[]) => void}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static on(event: string, callback: (...args: any[]) => void) {
    if (!NeonEventBus.listeners[event]) {
      NeonEventBus.listeners[event] = [];
    }
    NeonEventBus.listeners[event].push(callback);
  }

  /**
   * Stop listening to event bus events
   * @param event {string}
   * @param callback {(...args: any[]) => void}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static off(event: string, callback: (...args: any[]) => void) {
    NeonEventBus.listeners[event] = NeonEventBus.listeners[event].filter((fn) => fn !== callback);
  }

  /**
   * Emit event bus events
   * @param event {string}
   * @param args {any[]}
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static emit(event: string, ...args: any[]) {
    NeonEventBus.listeners[event]?.forEach((callbackFn) => callbackFn(...args));
  }
}
