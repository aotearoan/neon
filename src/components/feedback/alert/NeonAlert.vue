<template>
  <div class="neon-alert__wrapper">
    <neon-alert-container v-model="topLeft" placement="top-left" />
    <neon-alert-container v-model="topRight" placement="top-right" />
    <neon-alert-container v-model="bottomLeft" placement="bottom-left" />
    <neon-alert-container v-model="bottomRight" placement="bottom-right" />
    <neon-toast-container v-model="top" placement="top" />
    <neon-toast-container v-model="bottom" placement="bottom" />
  </div>
</template>

<script lang="ts" src="./NeonAlert.ts"></script>
