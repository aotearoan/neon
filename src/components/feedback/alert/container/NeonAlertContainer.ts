import Vue from 'vue';
import { Component, Prop } from 'vue-property-decorator';
import { NeonAlertLevel } from '../../../../common/enums/NeonAlertLevel';
import { NeonAlertPlacement } from '../../../../common/enums/NeonAlertPlacement';
import NeonIcon from '../../../presentation/icon/NeonIcon.vue';
import NeonLink from '../../../navigation/link/NeonLink.vue';
import { NeonAlertModel } from '../NeonAlertModel';

/**
 * This is an internal component for rendering alerts.
 */
@Component({
  components: {
    NeonIcon,
    NeonLink,
  },
})
export default class NeonAlertContainer extends Vue {
  @Prop({ required: true })
  private value!: NeonAlertModel[];

  @Prop({ required: true })
  private placement!: NeonAlertPlacement;

  private closeMessage(id: number) {
    this.$emit(
      'input',
      this.value.filter((msg) => msg.id !== id),
    );
  }

  private icon(level: NeonAlertLevel) {
    switch (level) {
      case NeonAlertLevel.Info:
        return 'info-circle';
      case NeonAlertLevel.Success:
        return 'check-circle';
      case NeonAlertLevel.Warn:
        return 'exclamation-circle';
      case NeonAlertLevel.Error:
        return 'times-circle';
    }
  }
}
