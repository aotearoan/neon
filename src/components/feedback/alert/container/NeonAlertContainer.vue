<template>
  <div :class="`neon-alert--${placement}`" class="neon-alert">
    <transition-group name="neon-alert-transition">
      <div
        v-for="(message, index) in modelValue"
        :key="message.id"
        :aria-describedby="message.message ? `alertMessage${index}` : undefined"
        :aria-labelledby="message.title ? `alertTitle${index}` : undefined"
        :class="[
          `neon-alert__message--${message.level}`,
          {
            'neon-alert__message--dismissible': message.dismissible,
            'neon-alert__message--with-actions': message.primaryAction,
          },
        ]"
        :role="message.primaryAction ? 'dialog' : message.duration && message.duration > 0 ? 'alert' : 'alertdialog'"
        class="neon-alert__message"
        @click="!message.primaryAction && message.dismissible && closeMessage(message.id)"
      >
        <neon-icon :color="message.level" :name="icon(message.level)" />
        <div
          :role="!message.primaryAction ? 'button' : undefined"
          :tabindex="message.primaryAction ? -1 : 0"
          class="neon-alert__content"
          @keydown.space.prevent="!message.primaryAction && message.dismissible && closeMessage(message.id)"
          @keydown.enter="!message.primaryAction && message.dismissible && closeMessage(message.id)"
        >
          <div v-if="message.title" :id="`alertTitle${index}`" class="neon-alert__title">{{ message.title }}</div>
          <div v-if="message.message" :id="`alertMessage${index}`" class="neon-alert__body">{{ message.message }}</div>
        </div>
        <div v-if="message.primaryAction" class="neon-alert__actions">
          <neon-link
            v-if="message.primaryAction"
            class="neon-alert__action"
            outline-style="none"
            role="button"
            tabindex="0"
            @click="
              message.primaryAction.callback();
              closeMessage(message.id);
            "
            @keydown.space.prevent="
              message.primaryAction.callback();
              closeMessage(message.id);
            "
          >
            <span class="neon-alert__action-label">{{ message.primaryAction.label }}</span>
          </neon-link>
          <neon-link
            v-if="message.secondaryAction"
            class="neon-alert__action"
            outline-style="none"
            role="button"
            tabindex="0"
            @click="
              message.secondaryAction.callback();
              closeMessage(message.id);
            "
            @keydown.space.prevent="
              message.secondaryAction.callback();
              closeMessage(message.id);
            "
          >
            <span class="neon-alert__action-label">{{ message.secondaryAction.label }}</span>
          </neon-link>
        </div>
      </div>
    </transition-group>
  </div>
</template>

<script lang="ts" src="./NeonAlertContainer.ts"></script>
