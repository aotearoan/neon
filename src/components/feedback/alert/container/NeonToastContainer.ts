import { defineComponent } from 'vue';
import NeonIcon from '@/components/presentation/icon/NeonIcon.vue';
import type { NeonToastModel } from '../NeonToastModel';
import type { NeonVerticalPosition } from '@/common/enums/NeonVerticalPosition';

/**
 * This is an internal component for rendering alerts.
 */
export default defineComponent({
  name: 'NeonToastContainer',
  components: {
    NeonIcon,
  },
  props: {
    modelValue: { type: Object as () => Array<NeonToastModel>, required: true },
    placement: { type: String as () => NeonVerticalPosition, required: true },
  },
  setup(props, { emit }) {
    const closeMessage = (id: number) => {
      emit(
        'update:modelValue',
        props.modelValue.filter((msg) => msg.id !== id),
      );
    };

    return {
      closeMessage,
    };
  },
});
