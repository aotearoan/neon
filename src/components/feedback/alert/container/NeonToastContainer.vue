<template>
  <div :class="`neon-alert--${placement}`" class="neon-alert">
    <transition-group name="neon-alert-transition">
      <div
        v-for="(message, index) in modelValue"
        :key="message.id"
        :aria-labelledby="`toastTitle${index}`"
        :class="[`neon-toast__message--${message.level}`, { 'neon-toast__message--dismissible': message.dismissible }]"
        :role="message.duration > 0 ? 'alert' : 'alertdialog'"
        class="neon-toast__message"
        @click="message.dismissible && closeMessage(message.id)"
      >
        <div v-if="message.title" :id="`toastTitle${index}`" class="neon-toast__title">{{ message.title }}</div>
        <neon-icon
          v-if="message.dismissible"
          :role="message.dismissible ? 'button' : undefined"
          class="neon-toast__close"
          name="times"
          tabindex="0"
          @click="message.dismissible && closeMessage(message.id)"
          @keydown.space="message.dismissible && closeMessage(message.id)"
          @keydown.enter="message.dismissible && closeMessage(message.id)"
          @keypress.space.prevent=""
        />
      </div>
    </transition-group>
  </div>
</template>

<script lang="ts" src="./NeonToastContainer.ts"></script>
