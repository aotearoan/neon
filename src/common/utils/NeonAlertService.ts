import { NeonEventBus } from './NeonEventBus';
import type { NeonAlertMessage } from '../models/NeonAlertMessage';
import { NeonAlertLevel } from '../enums/NeonAlertLevel';

/**
 * NeonAlertService is a service for sending alerts to the <strong>NeonAlert</strong> component for display to the user.
 */
export class NeonAlertService {
  /**
   * Send an info message.
   *
   * @param alert The info message to display.
   */
  public static info(alert: NeonAlertMessage) {
    NeonAlertService.emit(NeonAlertLevel.Info, alert);
  }

  /**
   * Send a success message.
   *
   * @param alert The success message to display.
   */
  public static success(alert: NeonAlertMessage) {
    NeonAlertService.emit(NeonAlertLevel.Success, alert);
  }

  /**
   * Send a warning message.
   *
   * @param alert The warning message to display.
   */
  public static warn(alert: NeonAlertMessage) {
    NeonAlertService.emit(NeonAlertLevel.Warn, alert);
  }

  /**
   * Send an error message.
   *
   * @param alert The error message to display.
   */
  public static error(alert: NeonAlertMessage) {
    NeonAlertService.emit(NeonAlertLevel.Error, alert);
  }

  /**
   * Generate an event key so that all events are published on the correct topic.
   *
   * @param eventType Alert level of the event.
   *
   * @returns The event key for sending a message on <a href="/utils/NeonEventBus">NeonEventBus</a>.
   */
  public static generateEventKey(eventType: NeonAlertLevel) {
    return `${NeonEventBus.messagePrefix}-alert-${eventType}`;
  }

  private static emit(eventType: NeonAlertLevel, alert: NeonAlertMessage) {
    const eventKey = NeonAlertService.generateEventKey(eventType);
    NeonEventBus.emit(eventKey, alert);
  }
}
