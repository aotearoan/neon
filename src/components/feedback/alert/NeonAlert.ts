import { defineComponent, onMounted, onUnmounted, ref } from 'vue';
import { NeonEventBus } from '../../../common/utils/NeonEventBus';
import type { NeonAlertMessage } from '../../../common/models/NeonAlertMessage';
import { NeonAlertLevel } from '../../../common/enums/NeonAlertLevel';
import { NeonAlertService } from '../../../common/utils/NeonAlertService';
import NeonAlertContainer from './container/NeonAlertContainer.vue';
import NeonToastContainer from './container/NeonToastContainer.vue';
import type { NeonAlertModel } from './NeonAlertModel';
import type { NeonToastModel } from './NeonToastModel';
import type { NeonToastMessage } from '../../../common/models/NeonToastMessage';
import { NeonToastService } from '../../../common/utils/NeonToastService';
import { NeonAlertPlacement } from '../../../common/enums/NeonAlertPlacement';
import { NeonVerticalPosition } from '../../../common/enums/NeonVerticalPosition';

/**
 * NeonAlert is a component for presenting temporary notifications to the user. Place the component once inside your app
 * and use <strong>NeonAlertService</strong> to send events to the component.
 */
export default defineComponent({
  name: 'NeonAlert',
  components: {
    NeonAlertContainer,
    NeonToastContainer,
  },
  props: {
    /**
     * Whether alert messages are dismissible by clicking on them. This can also be set per alert message.
     */
    dismissible: { type: Boolean, default: true },
    /**
     * Duration to display messages before removing them. This can also be set per alert message. Set the duration to 0 to
     * disable the duration timer, NOTE: this requires the user to dismiss the alert by clicking on it.
     */
    duration: { type: Number, default: 2500 },
  },
  setup(props) {

    // alerts
    const topLeft = ref<Array<NeonAlertModel>>([]);
    const topRight = ref<Array<NeonAlertModel>>([]);
    const bottomLeft = ref<Array<NeonAlertModel>>([]);
    const bottomRight = ref<Array<NeonAlertModel>>([]);
    // toasts
    const top = ref<Array<NeonToastModel>>([]);
    const bottom = ref<Array<NeonToastModel>>([]);

    const internalId = ref(1);

    const removeAlert = (message: NeonAlertModel) => {
      switch (message.placement || NeonAlertPlacement.TopRight) {
        case NeonAlertPlacement.TopLeft:
          topLeft.value = topLeft.value.filter((m) => m.id !== message.id);
          break;
        case NeonAlertPlacement.TopRight:
          topRight.value = topRight.value.filter((m) => m.id !== message.id);
          break;
        case NeonAlertPlacement.BottomLeft:
          bottomLeft.value = bottomLeft.value.filter((m) => m.id !== message.id);
          break;
        case NeonAlertPlacement.BottomRight:
          bottomRight.value = bottomRight.value.filter((m) => m.id !== message.id);
          break;
      }
    };

    const enqueueAlert = (level: NeonAlertLevel, alert: NeonAlertMessage) => {
      const id = internalId.value;
      internalId.value = internalId.value + 1;
      const message: NeonAlertModel = {
        dismissible: props.dismissible,
        ...alert,
        level,
        id,
      };

      switch (alert.placement || NeonAlertPlacement.TopRight) {
        case NeonAlertPlacement.TopLeft:
          topLeft.value.unshift(message);
          break;
        case NeonAlertPlacement.TopRight:
          topRight.value.unshift(message);
          break;
        case NeonAlertPlacement.BottomLeft:
          bottomLeft.value.push(message);
          break;
        case NeonAlertPlacement.BottomRight:
          bottomRight.value.push(message);
          break;
      }

      const duration = alert.duration === undefined ? props.duration : alert.duration;
      if (duration > 0 && !alert.primaryAction) {
        setTimeout(() => removeAlert(message), duration);
      }
    };

    const removeToast = (message: NeonToastModel) => {
      switch (message.placement || NeonVerticalPosition.Top) {
        case NeonVerticalPosition.Top:
          top.value = top.value.filter((m) => m.id !== message.id);
          break;
        case NeonVerticalPosition.Bottom:
          bottom.value = bottom.value.filter((m) => m.id !== message.id);
          break;
      }
    };

    const enqueueToast = (level: NeonAlertLevel, toast: NeonToastMessage) => {
      const id = internalId.value;
      internalId.value = internalId.value + 1;
      const message: NeonToastModel = {
        dismissible: props.dismissible,
        ...toast,
        level,
        id,
      };

      switch (toast.placement || NeonVerticalPosition.Top) {
        case NeonVerticalPosition.Top:
          top.value.unshift(message);
          break;
        case NeonVerticalPosition.Bottom:
          bottom.value.push(message);
          break;
      }

      const duration = toast.duration === undefined ? props.duration : toast.duration;
      if (duration > 0) {
        setTimeout(() => removeToast(message), duration);
      }
    };

    const onInfoAlert = (alert: NeonAlertMessage) => {
      enqueueAlert(NeonAlertLevel.Info, alert);
    };

    const onSuccessAlert = (alert: NeonAlertMessage) => {
      enqueueAlert(NeonAlertLevel.Success, alert);
    };

    const onWarnAlert = (alert: NeonAlertMessage) => {
      enqueueAlert(NeonAlertLevel.Warn, alert);
    };

    const onErrorAlert = (alert: NeonAlertMessage) => {
      enqueueAlert(NeonAlertLevel.Error, alert);
    };

    const onInfoToast = (toast: NeonToastMessage) => {
      enqueueToast(NeonAlertLevel.Info, toast);
    };

    const onSuccessToast = (toast: NeonToastMessage) => {
      enqueueToast(NeonAlertLevel.Success, toast);
    };

    const onWarnToast = (toast: NeonToastMessage) => {
      enqueueToast(NeonAlertLevel.Warn, toast);
    };

    const onErrorToast = (toast: NeonToastMessage) => {
      enqueueToast(NeonAlertLevel.Error, toast);
    };

    onMounted(() => {
      // alerts
      NeonEventBus.on(NeonAlertService.generateEventKey(NeonAlertLevel.Info), onInfoAlert);
      NeonEventBus.on(NeonAlertService.generateEventKey(NeonAlertLevel.Success), onSuccessAlert);
      NeonEventBus.on(NeonAlertService.generateEventKey(NeonAlertLevel.Warn), onWarnAlert);
      NeonEventBus.on(NeonAlertService.generateEventKey(NeonAlertLevel.Error), onErrorAlert);
      // toast
      NeonEventBus.on(NeonToastService.generateEventKey(NeonAlertLevel.Info), onInfoToast);
      NeonEventBus.on(NeonToastService.generateEventKey(NeonAlertLevel.Success), onSuccessToast);
      NeonEventBus.on(NeonToastService.generateEventKey(NeonAlertLevel.Warn), onWarnToast);
      NeonEventBus.on(NeonToastService.generateEventKey(NeonAlertLevel.Error), onErrorToast);
    });

    onUnmounted(() => {
      // alerts
      NeonEventBus.off(NeonAlertService.generateEventKey(NeonAlertLevel.Info), onInfoAlert);
      NeonEventBus.off(NeonAlertService.generateEventKey(NeonAlertLevel.Success), onSuccessAlert);
      NeonEventBus.off(NeonAlertService.generateEventKey(NeonAlertLevel.Warn), onWarnAlert);
      NeonEventBus.off(NeonAlertService.generateEventKey(NeonAlertLevel.Error), onErrorAlert);
      // toast
      NeonEventBus.off(NeonToastService.generateEventKey(NeonAlertLevel.Info), onInfoToast);
      NeonEventBus.off(NeonToastService.generateEventKey(NeonAlertLevel.Success), onSuccessToast);
      NeonEventBus.off(NeonToastService.generateEventKey(NeonAlertLevel.Warn), onWarnToast);
      NeonEventBus.off(NeonToastService.generateEventKey(NeonAlertLevel.Error), onErrorToast);
    });

    return {
      topLeft,
      topRight,
      bottomLeft,
      bottomRight,
      top,
      bottom,
    };
  },
});
