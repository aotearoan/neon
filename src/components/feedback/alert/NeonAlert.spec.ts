import Vue from 'vue';
import { mount } from '@vue/test-utils';
import NeonAlert from './NeonAlert.vue';
import NeonAlertContainer from './container/NeonAlertContainer.vue';
import NeonToastContainer from './container/NeonToastContainer.vue';
import { NeonAlertService } from '../../../common/utils/NeonAlertService';
import { NeonToastService } from '../../../common/utils/NeonToastService';
import { NeonAlertPlacement } from '../../../common/enums/NeonAlertPlacement';
import { NeonVerticalPosition } from '../../../common/enums/NeonVerticalPosition';
import { NeonEventBus } from '../../../common/utils/NeonEventBus';

Vue.component('NeonAlertContainer', NeonAlertContainer);
Vue.component('NeonToastContainer', NeonToastContainer);

describe('NeonAlert', () => {
  it('renders info alert', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders success alert', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.success({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--success').element).toBeDefined();
      done();
    });
  });

  it('renders warn alert', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.warn({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--warn').element).toBeDefined();
      done();
    });
  });

  it('renders error alert', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.error({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--error').element).toBeDefined();
      done();
    });
  });

  it('renders info toast', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders success toast', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.success({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--success').element).toBeDefined();
      done();
    });
  });

  it('renders warn toast', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.warn({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--warn').element).toBeDefined();
      done();
    });
  });

  it('renders error toast', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.error({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--error').element).toBeDefined();
      done();
    });
  });

  it('renders top right alert placement', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--top-right .neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('overrides default placement', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test', placement: NeonAlertPlacement.TopLeft });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--top-left .neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders alert bottom left', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test', placement: NeonAlertPlacement.BottomLeft });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--bottom-left .neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders alert bottom right', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test', placement: NeonAlertPlacement.BottomRight });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--bottom-right .neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders toast top', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--top .neon-toast__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders toast bottom', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.info({ title: 'test', placement: NeonVerticalPosition.Bottom });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert--bottom .neon-toast__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders duration 0', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: { duration: 0 },
    });
    // when
    NeonAlertService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders message duration 0', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonAlertService.info({ title: 'test', duration: 0 });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-alert__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders toast duration 0', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: { duration: 0 },
    });
    // when
    NeonToastService.info({ title: 'test' });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--info').element).toBeDefined();
      done();
    });
  });

  it('renders toast message duration 0', (done) => {
    // given
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    NeonToastService.info({ title: 'test', duration: 0 });
    // then
    setTimeout(() => {
      expect(wrapper.find('.neon-toast__message--info').element).toBeDefined();
      done();
    });
  });

  it('removes listeners on destroy', () => {
    // given
    const offFn = NeonEventBus.off;
    NeonEventBus.off = jest.fn();
    const wrapper = mount(NeonAlert, {
      propsData: {},
    });
    // when
    wrapper.destroy();
    // then
    expect(NeonEventBus.off).toHaveBeenCalledTimes(8);
    NeonEventBus.off = offFn;
  });
});
