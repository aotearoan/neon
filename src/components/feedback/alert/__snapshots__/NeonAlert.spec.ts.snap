// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonAlert matches snapshot 1`] = `
<div class="neon-alert__wrapper" dismissable="true">
  <div class="neon-alert--top-left neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
  <div class="neon-alert--top-right neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
  <div class="neon-alert--bottom-left neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
  <div class="neon-alert--bottom-right neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
  <div class="neon-alert--top neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
  <div class="neon-alert--bottom neon-alert">
    <transition-group-stub></transition-group-stub>
  </div>
</div>
`;
